src/core/server/metrics/metrics_service.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { ReplaySubject } from 'rxjs';
+import { first, shareReplay } from 'rxjs/operators';
+import { CoreService } from '../../types';
+import { CoreContext } from '../core_context';
+import { Logger } from '../logging';
+import { InternalHttpServiceSetup } from '../http';
+import { InternalMetricsServiceSetup, InternalMetricsServiceStart, OpsMetrics } from './types';
+import { OpsMetricsCollector } from './ops_metrics_collector';
+import { opsConfig, OpsConfigType } from './ops_config';
+
+interface MetricsServiceSetupDeps {
+  http: InternalHttpServiceSetup;
+}
+
+/** @internal */
+export class MetricsService
+  implements CoreService<InternalMetricsServiceSetup, InternalMetricsServiceStart> {
+  private readonly logger: Logger;
+  private metricsCollector?: OpsMetricsCollector;
+  private collectInterval?: NodeJS.Timeout;
+  private metrics$ = new ReplaySubject<OpsMetrics>(1);
+
+  constructor(private readonly coreContext: CoreContext) {
+    this.logger = coreContext.logger.get('metrics');
+  }
+
+  public async setup({ http }: MetricsServiceSetupDeps): Promise<InternalMetricsServiceSetup> {
+    this.metricsCollector = new OpsMetricsCollector(http.server);
+
+    const metricsObservable = this.metrics$.pipe(shareReplay(1));
+
+    return {
+      getOpsMetrics$: () => metricsObservable,
+    };
```

Comment:
I exposed this in the `setup` API, even if the observable doesn't emit until core's `start` phase. Tell me if we prefer moving it to the `start` API instead.

Reply:
I think it makes sense if we implement lazy logic

Reply:
Makes sense to me to maintain the pattern of "register things in setup"